sw/device/lib/crypto/impl/ecc/ecdsa_p256.c

```diff
@@ -33,111 +34,113 @@ static const otbn_addr_t kOtbnVarEcdsaD0 = OTBN_ADDR_T_INIT(p256_ecdsa, d0);
 static const otbn_addr_t kOtbnVarEcdsaD1 = OTBN_ADDR_T_INIT(p256_ecdsa, d1);
 static const otbn_addr_t kOtbnVarEcdsaXr = OTBN_ADDR_T_INIT(p256_ecdsa, x_r);
 
-/* Mode is represented by a single word, 1 for sign and 2 for verify */
-static const uint32_t kOtbnEcdsaModeNumWords = 1;
-static const uint32_t kOtbnEcdsaModeSign = 1;
-static const uint32_t kOtbnEcdsaModeVerify = 2;
+/* Mode is represented by a single word. See `p256_ecdsa.s` for values. */
+static const uint32_t kOtbnEcdsaModeWords = 1;
+static const uint32_t kOtbnEcdsaModeKeygen = 0x3d4;
+static const uint32_t kOtbnEcdsaModeSign = 0x15b;
+static const uint32_t kOtbnEcdsaModeVerify = 0x727;
 
-// TODO: This implementation waits while OTBN is processing; it should be
-// modified to be non-blocking.
-status_t ecdsa_p256_sign(const ecdsa_p256_message_digest_t *digest,
-                         const ecdsa_p256_private_key_t *private_key,
-                         ecdsa_p256_signature_t *result) {
+status_t ecdsa_p256_keygen_start(void) {
   // Load the ECDSA/P-256 app. Fails if OTBN is non-idle.
   HARDENED_TRY(otbn_load_app(kOtbnAppEcdsa));
 
-  // Set mode so start() will jump into p256_ecdsa_sign.
-  HARDENED_TRY(otbn_dmem_write(kOtbnEcdsaModeNumWords, &kOtbnEcdsaModeSign,
+  // Set mode so start() will jump into keygen.
+  HARDENED_TRY(otbn_dmem_write(kOtbnEcdsaModeWords, &kOtbnEcdsaModeKeygen,
+                               kOtbnVarEcdsaMode));
+
+  // Start the OTBN routine.
+  return otbn_execute();
+}
+
+status_t ecdsa_p256_keygen_finalize(p256_masked_scalar_t *private_key,
+                                    p256_point_t *public_key) {
+  // Spin here waiting for OTBN to complete.
+  HARDENED_TRY(otbn_busy_wait_for_done());
```

Comment:
Should this return a busy status to make the `*_finalize` calls exit early if otbn is busy? This way the caller can opportunistically retry later.

Reply:
Maybe a better way would be to expose some kind of `check_async_status()` in the top-level API that tells the caller if OTBN is busy or not? Then it could be used for all async operations, and would preserve the convenience of being able to simply call `start()` and `finalize()` in sequence if the caller is OK with blocking.

Reply:
`check_async_status()` sounds good to me.